### ISSUE TITLE
New CMS: Improve hashing of BigInt

### ISSUE THREAD

**Issue by miguno** [twitter/algebird#392]
## Summary

The latest implementation of CMS ships with a default hashing implementation for `BigInt` that is based on the previous hashing approach for `Long`.  Unfortunately, this implementation for `BigInt` is "broken" in my opinion.

I think I have found a clean fix, which I will send in a separate pull request.  It's actually only a few lines of code, and the reason I am writing this ticket is simply to justify these lines and solicit peer review. :-)

## What is broken in CMS[K] for BigInt's?

**TL;DR: The current hashing causes more collisions "than expected", leading to counter-intuitive CMS counting behavior because similar but different keys end up being treated the same.  For instance, "twitter.com" is hashed to the same counting table slot as "www.twitter.com" and "mail.twitter.com".**

Currently we use the following hashing function for `BigInt`:

```scala
  implicit object CMSHasherBigInt extends CMSHasher[BigInt] {

    def hash(a: Int, b: Int, width: Int)(x: BigInt) = {
      val unModded: BigInt = (x * a) + b
      val modded: BigInt = (unModded + (unModded >> 32)) & PRIME_MODULUS
      val h = modded.toInt % width
      assert(h >= 0, "hash must not be negative")
      h
    }

  }
```

I will now show a few examples that illustrate what is wrong with the current hashing implementation for `BigInt`.

```
$ ./sbt "project algebird-core" console
```
```scala
scala> import com.twitter.algebird._
scala> import com.twitter.algebird.CMSHasherImplicits._
scala> val a = 5; val b = 0; val width = 999;

// Example: We hash domain names.
// Result: Different domain names are assigned to the same slot.
Seq("twitter.com", "www.twitter.com", "mail.twitter.com").map { s => CMSHasherBigInt.hash(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res133: Seq[Int] = List(889, 889, 889)   //  <<< again, all inputs hash to the same slot!

// Let's see how bad it is.  Ok, once we have only "ter.com" to hash we get a different slot.
scala> Seq("twitter.com", "witter.com", "itter.com", "tter.com", "ter.com").map { s => CMSHasherBigInt.hash(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res209: Seq[Int] = List(889, 889, 889, 889, 490)

// Also, I don't think that these domains should be considered the same...
scala> Seq("twitter.com", "litter.com", "bitter.com", "sitter.com").map { s => CMSHasherBigInt.hash(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res212: Seq[Int] = List(889, 889, 889, 889)

//
// Let's try to pinpoint the issue.
//

// Observation 1: The hash function ignores all but the least significant (right-most) bits.
// This means that, say, similar but different usernames may be considered as identical by CMS.

// Example: We hash "A","AA", "AAA", etc.
// Result: All `A`-strings with length >= 8 are assigned to the same slot.  Oops.
scala> (1 to 20).map { rep => "A" * rep } map { s => CMSHasherBigInt.hash(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res174: scala.collection.immutable.IndexedSeq[Int] = Vector(325, 774, 66, 304, 629, 81, 370, 124, 124, 124, 124, 124, 124, 124, 124, 124, 124, 124, 124, 124)

// Example: We vary the first character of sufficiently long strings, i.e. "Axxxx...", "Bxxxx...".
// Results: For all padLengths >= 8 the strings are assigned to the same slot.  Oops.
scala> val padLength = 8
scala> ('A' to 'Z') map (_.toString) map { s => s + "x" * padLength } map { s => CMSHasherBigInt.hash(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res196: scala.collection.immutable.IndexedSeq[Int] = Vector(113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113)
scala> val padLength = 12
scala> ('A' to 'Z') map (_.toString) map { s => s + "x" * padLength } map { s => CMSHasherBigInt.hash(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res200: scala.collection.immutable.IndexedSeq[Int] = Vector(113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113, 113)

// Observation 2: Non-random hashes lead to predictable slot patterns (think: geometric series).
// Note how in these two examples the slots always increase by a fixed value (here: `+5`).
// This is expected given the current hash function, but still this is a property we do not like.
scala> Seq(1, 2, 3, 4, 5).map { i => CMSHasherBigInt.hash(a, b, width)(BigInt(i)) }
res150: Seq[Int] = List(5, 10, 15, 20, 25)

// Same for the Long-based hashing, of course.
scala> Seq(1, 2, 3, 4, 5).map { i => CMSHasherLong.hash(a, b, width)(i) }
res222: Seq[Int] = List(5, 10, 15, 20, 25)
```

## How to fix

_One might argue that an Algebird user could of course opt to implement her own `BigInt` hashing function.  However, I'd say that the current default `BigInt` hashing is so counter-intuitive to the general user that I'd consider it broken, and thus we should fix it at the source._

### Requirements

We need a hash function that:

1. spreads the input data randomly across the available `width` slots.
2. if the inputs `a` and `b` are "close" to each other, then their hash values should be far apart.
3. is not computationally expensive, which rules out cryptographic hash functions like MD5 or SHA1.

### Proposed solution: Use Murmur3 for hashing BigInt

**TL;DR: Murmur3 seems to be a good choice, although other options may be valid, too.  Experiments and benchmarks confirm the applicability of Murmur3.**

I did a cursory evaluation of the following hash functions:

- Murmur3
- Jenkins (e.g. implemented in [twitter.util.Hashing](https://github.com/twitter/util/blob/master/util-hashing/src/main/scala/com/twitter/hashing/Hashable.scala))
- [SipHash](https://131002.net/siphash/), CityHash
- CRC32 and similar error detecting codes
- And actually a handful of others that I forgot in the meantime unfortunately (blame Xmas holidays!)

> I also came across [this benchmark](https://docs.google.com/spreadsheets/d/1_q2EVcxA2HjcrlVMbaqXwMj31h9M5-Bqj_m8vITOwwk/edit?pli=1#gid=0) of the various hashes shipped with Google Guava.  Note: In my benchmarking Guava's Murmur3 with 128-bit hashes turned out to be actually faster than Guava's 32-bit version of Murmur3.

Eventually I reached the conclusion that Murmur3 seems to be the best option:

- First, Murmur2/Murmur3 [has been shown](http://programmers.stackexchange.com/questions/49550/which-hashing-algorithm-is-best-for-uniqueness-and-speed) to perform well in a random distribution of regular keys.
- In my benchmarking (see below) Murmur3 ran as fast -- even slightly faster -- than the current, broken hash implementation.
- Scala ships with a built-in implementation of Murmur3 that outputs 32-bit hashes, so we wouldn't need to introduce a code dependency to Algebird just because of the `BigInt` hashing.

The Murmur3-based CMS hashing function looks as follows:

```scala
def murmurHashScala(a: Int, b: Int, width: Int)(x: BigInt) = {
    val hash: Int = scala.util.hashing.MurmurHash3.arrayHash(x.toByteArray, a)
    val h = {
      // We only want positive integers for the subsequent modulo.
      // This method mimics Java's Hashtable implementation, and it requires
      // `hash` to be an `Int` = have 32 bits (to match with `0x7FFFFFFF`).
      val positiveHash = hash & 0x7FFFFFFF
      positiveHash % width
    }
    assert(h >= 0, "hash must not be negative")
    h
  }
```

**Important notes:**

* The original CMS paper requires `d` (depth) pair-wise independent hash functions.  Because Murmur3 has "almost" cryptographic quality (see e.g. [this discussion on StackOverflow](http://stackoverflow.com/questions/12120313/generating-k-pairwise-independent-hash-functions)), I argue that we can use different _seeds_ for Murmur3 instead of creating a universal hashing family.  Here, we'd use the `a` parameter in CMS as the seed for Murmur3;  assuming this is not a stupid decision per se (and I might be mistaken), this would be an elegant solution because CMS currently ignores the `b` parameter anyway, so we only need to take `a` into account.

Here are the same examples as I showed above but this time using Murmur3 (with 32-bit hashes as output).  All results indicate that the CMS hashing behavior has improved and shows the expected behavior.

```scala
// Example: We hash "A","AA", "AAA", etc.
// Result: Good, all `A`-strings are well spread across the slots.
scala> (1 to 20).map { rep => "A" * rep } map { s => murmurHashScala(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res223: scala.collection.immutable.IndexedSeq[Int] = Vector(561, 636, 318, 639, 707, 736, 204, 186, 194, 426, 112, 827, 842, 143, 581, 874, 890, 492, 5, 679)

// Example: We vary the first character of sufficiently long strings, i.e. "Axxxx...", "Bxxxx...".
// Results: Good, all inputs are well spread across the slots.
scala> val padLength = 8
scala> ('A' to 'Z') map (_.toString) map { s => s + "x" * padLength } map { s => murmurHashScala(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res224: scala.collection.immutable.IndexedSeq[Int] = Vector(194, 869, 369, 784, 444, 472, 862, 342, 680, 698, 487, 114, 751, 665, 758, 574, 20, 734, 543, 721, 767, 689, 77, 299, 152, 808)
scala> val padLength = 12
scala> ('A' to 'Z') map (_.toString) map { s => s + "x" * padLength } map { s => murmurHashScala(a, b, width)(BigInt(s.getBytes("UTF8"))) }
res226: scala.collection.immutable.IndexedSeq[Int] = Vector(914, 256, 649, 983, 305, 453, 218, 598, 641, 729, 49, 653, 730, 77, 329, 944, 761, 348, 632, 456, 762, 804, 605, 563, 236, 202)

// Good: Hashes are actually random now, leading to unpredictable slot patterns.
scala> Seq(1, 2, 3, 4, 5).map { i => murmurHashScala(a, b, width)(BigInt(i)) }
res227: Seq[Int] = List(920, 215, 33, 439, 949)
```

### Benchmark results

* [Benchmark code: CMSHashingBenchmark.scala](https://gist.github.com/miguno/0fd6b36d52f477762514) -- I decided to put this into a gist because I do not want to include it in the pull request as this would require adding Google Guava as a dependency to the sbt build of Algebird.

Here:

* `BrokenCurrentHash` is the current BigInt hashing in CMS.
* `MurmurHashScala` uses Scala's built-in Murmur3 implementation.
* `MurmurHash{32,128}Guava` uses Google Guava's Murmur3 implementation with 32-bit and 128-bit hashes, respectively.
* `maxBits`: The size of the randomly generated `BigInt` instances used as test input data.
* `width`: Represents the width of the counting table (note that this benchmark only simulates a single row of the counting table).

Summary of results:

- Scala's built-in Murmur3 implementation was the fastest in this benchmark.
    - It was particularly faster for small `BigInt` instances (in the benchmark: 128 bits).
- Interestingly, Guava's Murmur3 implementation was faster when producing 128-bit hashes than when producing 32-bit hashing.

```
[info] maxBits                                  benchmark  width    ms linear runtime
[info]     128  BrokenCurrentHashWithRandomMaxBitsNumbers     11  34.7 ===
[info]     128  BrokenCurrentHashWithRandomMaxBitsNumbers    544  34.8 ===
[info]     128  BrokenCurrentHashWithRandomMaxBitsNumbers   2719  34.9 ===
[info]     128  BrokenCurrentHashWithRandomMaxBitsNumbers 271829  35.5 ===
[info]     128    MurmurHashScalaWithRandomMaxBitsNumbers     11  22.4 ==
[info]     128    MurmurHashScalaWithRandomMaxBitsNumbers    544  22.5 ==
[info]     128    MurmurHashScalaWithRandomMaxBitsNumbers   2719  21.9 ==
[info]     128    MurmurHashScalaWithRandomMaxBitsNumbers 271829  22.4 ==
[info]     128  MurmurHash32GuavaWithRandomMaxBitsNumbers     11  33.8 ===
[info]     128  MurmurHash32GuavaWithRandomMaxBitsNumbers    544  37.1 ===
[info]     128  MurmurHash32GuavaWithRandomMaxBitsNumbers   2719  40.2 ====
[info]     128  MurmurHash32GuavaWithRandomMaxBitsNumbers 271829  42.2 ====
[info]     128 MurmurHash128GuavaWithRandomMaxBitsNumbers     11  35.9 ===
[info]     128 MurmurHash128GuavaWithRandomMaxBitsNumbers    544  39.3 ====
[info]     128 MurmurHash128GuavaWithRandomMaxBitsNumbers   2719  41.1 ====
[info]     128 MurmurHash128GuavaWithRandomMaxBitsNumbers 271829  43.8 ====
[info]    1024  BrokenCurrentHashWithRandomMaxBitsNumbers     11 137.2 ==============
[info]    1024  BrokenCurrentHashWithRandomMaxBitsNumbers    544 133.8 ==============
[info]    1024  BrokenCurrentHashWithRandomMaxBitsNumbers   2719 142.7 ===============
[info]    1024  BrokenCurrentHashWithRandomMaxBitsNumbers 271829 139.0 ==============
[info]    1024    MurmurHashScalaWithRandomMaxBitsNumbers     11 136.5 ==============
[info]    1024    MurmurHashScalaWithRandomMaxBitsNumbers    544 137.2 ==============
[info]    1024    MurmurHashScalaWithRandomMaxBitsNumbers   2719 146.7 ===============
[info]    1024    MurmurHashScalaWithRandomMaxBitsNumbers 271829 142.6 ===============
[info]    1024  MurmurHash32GuavaWithRandomMaxBitsNumbers     11 149.8 ===============
[info]    1024  MurmurHash32GuavaWithRandomMaxBitsNumbers    544 157.2 ================
[info]    1024  MurmurHash32GuavaWithRandomMaxBitsNumbers   2719 157.7 ================
[info]    1024  MurmurHash32GuavaWithRandomMaxBitsNumbers 271829 158.5 ================
[info]    1024 MurmurHash128GuavaWithRandomMaxBitsNumbers     11 157.0 ================
[info]    1024 MurmurHash128GuavaWithRandomMaxBitsNumbers    544 158.6 ================
[info]    1024 MurmurHash128GuavaWithRandomMaxBitsNumbers   2719 164.4 =================
[info]    1024 MurmurHash128GuavaWithRandomMaxBitsNumbers 271829 157.7 ================
[info]    2048  BrokenCurrentHashWithRandomMaxBitsNumbers     11 239.8 =========================
[info]    2048  BrokenCurrentHashWithRandomMaxBitsNumbers    544 218.7 =======================
[info]    2048  BrokenCurrentHashWithRandomMaxBitsNumbers   2719 245.0 ==========================
[info]    2048  BrokenCurrentHashWithRandomMaxBitsNumbers 271829 244.8 ==========================
[info]    2048    MurmurHashScalaWithRandomMaxBitsNumbers     11 250.7 ==========================
[info]    2048    MurmurHashScalaWithRandomMaxBitsNumbers    544 253.7 ===========================
[info]    2048    MurmurHashScalaWithRandomMaxBitsNumbers   2719 252.7 ==========================
[info]    2048    MurmurHashScalaWithRandomMaxBitsNumbers 271829 252.2 ==========================
[info]    2048  MurmurHash32GuavaWithRandomMaxBitsNumbers     11 260.2 ===========================
[info]    2048  MurmurHash32GuavaWithRandomMaxBitsNumbers    544 281.8 ==============================
[info]    2048  MurmurHash32GuavaWithRandomMaxBitsNumbers   2719 265.0 ============================
[info]    2048  MurmurHash32GuavaWithRandomMaxBitsNumbers 271829 270.4 ============================
[info]    2048 MurmurHash128GuavaWithRandomMaxBitsNumbers     11 246.7 ==========================
[info]    2048 MurmurHash128GuavaWithRandomMaxBitsNumbers    544 247.8 ==========================
[info]    2048 MurmurHash128GuavaWithRandomMaxBitsNumbers   2719 257.9 ===========================
[info]    2048 MurmurHash128GuavaWithRandomMaxBitsNumbers 271829 260.7 ===========================
[info]
[info] vm: java
[info] trial: 0
[info] a: 5123456
[info] b: 0
[info] operations: 100000
```

Note: I did not profile the memory impacts of the various hashing options yet.

**Comment by johnynek** [comment]
Once again, an amazingly detailed and rigorous contribution.

This looks great to me. Thank you.

**Comment by miguno** [comment]
FWIW here are the benchmark results for the committed code.  No changes apart from removing the Google Guava variants for Murmur3 because I didn't want to introduce a dependency on Guava just for this benchmark.

`MurmurHashScala*` is the new hash function we use for `BigInt` in CMS.

```
[info] maxBits                                 benchmark  width    ms linear runtime
[info]     128 BrokenCurrentHashWithRandomMaxBitsNumbers     11  37.2 ====
[info]     128 BrokenCurrentHashWithRandomMaxBitsNumbers    544  35.9 ====
[info]     128 BrokenCurrentHashWithRandomMaxBitsNumbers   2719  37.2 ====
[info]     128 BrokenCurrentHashWithRandomMaxBitsNumbers 271829  36.1 ====
[info]     128   MurmurHashScalaWithRandomMaxBitsNumbers     11  23.1 ==
[info]     128   MurmurHashScalaWithRandomMaxBitsNumbers    544  23.4 ==
[info]     128   MurmurHashScalaWithRandomMaxBitsNumbers   2719  23.0 ==
[info]     128   MurmurHashScalaWithRandomMaxBitsNumbers 271829  22.7 ==
[info]    1024 BrokenCurrentHashWithRandomMaxBitsNumbers     11 140.7 ================
[info]    1024 BrokenCurrentHashWithRandomMaxBitsNumbers    544 139.2 ===============
[info]    1024 BrokenCurrentHashWithRandomMaxBitsNumbers   2719 141.4 ================
[info]    1024 BrokenCurrentHashWithRandomMaxBitsNumbers 271829 143.7 ================
[info]    1024   MurmurHashScalaWithRandomMaxBitsNumbers     11 134.9 ===============
[info]    1024   MurmurHashScalaWithRandomMaxBitsNumbers    544 145.5 ================
[info]    1024   MurmurHashScalaWithRandomMaxBitsNumbers   2719 144.9 ================
[info]    1024   MurmurHashScalaWithRandomMaxBitsNumbers 271829 150.6 =================
[info]    2048 BrokenCurrentHashWithRandomMaxBitsNumbers     11 251.9 ============================
[info]    2048 BrokenCurrentHashWithRandomMaxBitsNumbers    544 255.3 =============================
[info]    2048 BrokenCurrentHashWithRandomMaxBitsNumbers   2719 252.4 ============================
[info]    2048 BrokenCurrentHashWithRandomMaxBitsNumbers 271829 242.3 ===========================
[info]    2048   MurmurHashScalaWithRandomMaxBitsNumbers     11 256.4 =============================
[info]    2048   MurmurHashScalaWithRandomMaxBitsNumbers    544 255.9 =============================
[info]    2048   MurmurHashScalaWithRandomMaxBitsNumbers   2719 262.6 ==============================
[info]    2048   MurmurHashScalaWithRandomMaxBitsNumbers 271829 252.6 ============================
[info]
[info] vm: java
[info] trial: 0
[info] a: 5123456
[info] b: 0
[info] operations: 100000
```

**Comment by johnynek** [comment]
Thank you again for your careful and complete work. It is appreciated.
